tests/Unit/Http/Controller/FollowControllerTest.php

```diff
@@ -43,15 +45,89 @@ public function test_follow_success()
+        $this->activityMock->shouldReceive('create')->withAnyArgs($dataActivity)->andReturn(true).
+        $this->activityMock->shouldReceive('saveDataToActivity')->withAnyArgs($data, $dataActivity)->andReturn(true);
         $controller = $this->followController->follow($request);
+
         $result = [
             'user_id' => $user->id,
             'follow_id' => $follow->id,
         ];
         $this->assertEquals($result, $controller->original);
     }
 
+    public function test_follow_fails_by_getUserRelationFriend_fails()
+    {
+        $user = factory(User::class)->make();
+        $user->id = config('id.id_user_login_current');
+        $follow = factory(Follow::class)->make();
+        $follow->id = config('id.id_user_has_been_follow');
+        $this->be($user);
+        $request = new Request;
+        $data = [
+            'user_id' => $user->id,
+            'follow_id' => $follow->id,
+        ];
+        
+        $dataNoti = [
+            'message' => 'profile.action_follow_history',
+            'user_name' => 'hathu99',
+        ];
+        $dataActivity = [
+            'user_id' => $user->id,
+            'action' => 2,
+            'notify' => json_encode($dataNoti),
+        ];
+        $this->followMock->shouldReceive('getFollower')->with($request->id)->andReturn(true);
+        $this->followMock->shouldReceive('getUserRelationFriend')->with($request->id)->andReturn(true);
+        $this->followMock->shouldReceive('create')->withAnyArgs($data)->andReturn($follow);
+        $this->activityMock->shouldReceive('create')->withAnyArgs($dataActivity)->andReturn([]).
+        $this->activityMock->shouldReceive('saveDataToActivity')->withAnyArgs($data, $dataActivity)->andReturn(true);
+        $controller = $this->followController->follow($request);
+        $this->assertEquals($data, $controller->original);
+    }
+
+    public function test_follow_fail_by_create()
+    {
+        $user = factory(User::class)->make();
+        $user->id = config('id.id_user_login_current');
+        $follow = factory(Follow::class)->make();
+        $follow->id = config('id.id_user_has_been_follow');;
+        $this->be($user);
+        $request = new Request;
+        $data = [
+            'status' => config('id.status_fail_404'),
+            'message' => trans('mes.fail'),
+        ];
+        $dataNoti = [
+            'message' => 'profile.action_follow_history',
+            'user_name' => 'hathu99',
+        ];
+        $dataActivity = [
+            'user_id' => $user->id,
+            'action' => 2,
+            'notify' => json_encode($dataNoti),
+        ];
+        $this->followMock->shouldReceive('getFollower')->with($request->id)->andReturn(true);
+        $this->followMock->shouldReceive('getUserRelationFriend')->with($request->id)->andReturn(true);
+        $this->followMock->shouldReceive('create')->withAnyArgs($data)->andReturn(false);
```

Comment:
```suggestion
        $this->followMock->shouldReceive('create')->with($data)->andReturn(false);
```
Tại sao ở đây lại sử dụng **withAnyArgs()**? sao ko sử dụng with()?

Reply:
dùng with thì cũng được, cái withAnyArgs() nó nhận các kiểu giá trị truyền vào thôi.

Reply:
nếu đầu vào có thể xác định được thì nên dùng with để test chặt chẽ nhé.dùng  withAnyArgs thì nó là mặc định rồi, thêm vào thì thừa.

Reply:
trước a có dùng with nhưng truyền vào data là mảng thì nó ko nhận đc nên phải dùng withAnyArgs() ý.

Reply:
hàm **with** ko nhận tức là anh xác định đầu vào của hàm trong test sai
HOẶC
hàm trong controller truyền vào sai so với giá trị mong đợi.

Reply:
à ừ nó bị sai đầu vào, a fix lại rồi.